File: problems/645/solution.go
package main
func findErrorNums(nums []int) []int {
    dif, a, b := 0, 0, 0
    
    for _, num := range nums {
        dif ^= num
    }
    
    for i := 1; i <= len(nums); i++ {
        dif ^= i
    }
    
    dif = dif & -dif
    
    for _, num := range nums {
        if (dif & num) != 0 {
            a ^= num
        } else {
            b ^= num
        }
    }
    
    for i := 1; i <= len(nums); i++ {
        if (dif & i) != 0 {
            a ^= i
        } else {
            b ^= i
        }
    }
    
    for _, num := range nums {
        if num == b {
            return []int{b, a}
        }
    }
    
    return []int{a, b}
}

File: problems/645/solution.cpp
class Solution {
public:
    vector<int> findErrorNums(vector<int>& nums) {
        int dif = 0, a = 0, b = 0;
        
        for (const int& num: nums) {
            dif ^= num;
        }
        
        for (int i = 1; i <= nums.size(); i++) {
            dif ^= i;
        }
        
        dif = dif & -dif;
        
        for (const int& num: nums) {
            if (dif & num) {
                a ^= num;
            } else {
                b ^= num;
            }
        }
        
        for (int i = 1; i <= nums.size(); i++) {
            if (dif & i) {
                a ^= i;
            } else {
                b ^= i;
            }
        }
        
        for (const int& num: nums) {
            if (num == b) {
                return {b, a};
            }
        }
        
        return {a, b};
    }
    
    vector<int> findErrorNums2(vector<int>& nums) {
        vector<int> ans(2);
        
        for (int i = 0; i < nums.size(); i++) {
            int num = abs(nums[i]);
            
            if (nums[num - 1] > 0) {
                nums[num - 1] *= -1;
            } else {
                ans[0] = num;
            }
        }
        
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] > 0) {
                ans[1] = i + 1;
            }
        }
        
        return ans;
    }
    
    vector<int> findErrorNums1(vector<int>& nums) {
        vector<int> ans;
        unordered_map<int, int> m;
        
        for (const int& num: nums) {
            m[num]++;
            
            if (m[num] >= 2) {
                ans.push_back(num);
            }
        }
        
        for (int i = 1; i <= nums.size(); i++) {
            if (not m.count(i)) {
                ans.push_back(i);
                break;
            }
        }
        
        return ans;
    }
};